src/core/radio/radio.cpp

```diff
@@ -92,4 +95,128 @@ Error Radio::Transmit(Mac::TxFrame &aFrame)
 }
 #endif // OPENTHREAD_CONFIG_RADIO_LINK_IEEE_802_15_4_ENABLE
 
+#if OPENTHREAD_CONFIG_RADIO_STATS_ENABLE && (OPENTHREAD_FTD || OPENTHREAD_MTD)
+RadioStatistics::RadioStatistics(void)
+    : mStatus(kDisabled)
+    , mSleepTime(0)
+    , mRxTime(0)
+    , mTxTime(0)
+    , mLastCalTime(0)
+{
+}
+
+void RadioStatistics::ChangeStatus(Status aStatus)
+{
+    TimeMicro curTime = TimerMicro::GetNow();
+    if (mStatus != kDisabled)
+    {
+        UpdateTime(curTime);
+    }
+    else
+    {
+        mLastCalTime = TimerMicro::GetNow();
+    }
+
+    mStatus = aStatus;
+}
+
+void RadioStatistics::HandleReceiveAt(uint32_t aDurationUs)
+{
+    // The actual rx time of ReceiveAt cannot be obtained from software level. This is a workaround.
+    if (mStatus == kSleep)
+    {
+        mRxTime += aDurationUs;
+    }
+}
+
+void RadioStatistics::TxDone(otError aError, uint16_t aBytes)
+{
+    if (aError == kErrorNone || aError == kErrorNoAck)
+    {
+        uint32_t txTimeUs = (aBytes + Mac::Frame::kPhyHeaderSize) * Radio::kSymbolsPerOctet * Radio::kSymbolTime;
+        UpdateTime(TimerMicro::GetNow());
+        mTxTime += txTimeUs;
+        if (mStatus == kReceive)
+        {
+            if (mRxTime > txTimeUs)
+            {
+                mRxTime -= txTimeUs;
+            }
+            else
+            {
+                mRxTime = 0;
+            }
+        }
+        else if (mStatus == kSleep)
+        {
+            if (mSleepTime > txTimeUs)
+            {
+                mSleepTime -= txTimeUs;
+            }
+            else
+            {
+                mSleepTime = 0;
+            }
+        }
+    }
+}
+
+void RadioStatistics::RxDone(otError aError)
+{
+    uint32_t ackTimeUs;
+
+    VerifyOrExit(aError == kErrorNone);
+
+    UpdateTime(TimerMicro::GetNow());
+    ackTimeUs =
+        (Mac::Frame::kImmAckLength + Mac::Frame::kPhyHeaderSize) * Radio::kSymbolsPerOctet *
+        Radio::kSymbolTime; // Currently we cannot know the actual length of ACK. So assume the ACK is an immediate ACK.
+    mTxTime += ackTimeUs;
+    if (mStatus == kReceive)
+    {
+        mRxTime -= ackTimeUs;
+    }
+
+exit:
+    return;
+}
+
+void RadioStatistics::GetStats(otRadioStats &aRadioStats)
+{
+    TimeMicro curTime = TimerMicro::GetNow();
+
+    if (mStatus != kDisabled)
+    {
+        UpdateTime(curTime);
+    }
+
+    aRadioStats.mSleepTime = mSleepTime;
+    aRadioStats.mTxTime    = mTxTime;
+    aRadioStats.mRxTime    = mRxTime;
+}
+
+void RadioStatistics::ResetTime(void)
+{
+    mSleepTime   = 0;
+    mRxTime      = 0;
+    mTxTime      = 0;
+    mLastCalTime = TimerMicro::GetNow();
+}
+
+void RadioStatistics::UpdateTime(TimeMicro aCurTime)
```

Comment:
Could we remove the parameter `aCurTime ` and add `TimeMicro curTime = TimerMicro::GetNow();` in the function `UpdateTime ()`?